File: app/tests/support/hooks.ts
import { Before, After } from "@cucumber/cucumber";
import { chromium, request as pwRequest } from "playwright";
import { CustomWorld } from "./world";
import { BASE_URL } from "./config";
import fs from "fs";

Before({ tags: "@ui" }, async function (this: CustomWorld) {
  this.browser = await chromium.launch({ headless: false });
  this.context = await this.browser.newContext();
  this.page = await this.context.newPage();
});

Before({ tags: "@api" }, async function (this: CustomWorld) {
  this.request = await pwRequest.newContext({
    baseURL: BASE_URL,
    extraHTTPHeaders: { "Content-Type": "application/json" }
  });
});

After({ tags: "@ui" }, async function (this: CustomWorld, scenario) {
  try {
    if (scenario.result?.status === "FAILED" && this.page) {
      fs.mkdirSync("tests/screenshots", { recursive: true });
      await this.page.screenshot({
        path: `tests/screenshots/${Date.now()}-FAILED.png`,
        fullPage: true
      });
    }
  } finally {
    await this.context?.close();
    await this.browser?.close();
  }
});

After({ tags: "@api" }, async function (this: CustomWorld) {
  await this.request?.dispose();
});

File: Test/src/test/java/utils/Hooks.java
package support;

import com.microsoft.playwright.*;
import io.cucumber.java.*;

import java.nio.file.*;
import java.time.Instant;

public class Hooks {
    private final TestContext ctx;

    public Hooks(TestContext ctx) {
        this.ctx = ctx;
    }
    @Before("@ui")
    public void beforeUi() {
        ctx.playwright = Playwright.create();
        ctx.browser = ctx.playwright.chromium().launch(new BrowserType.LaunchOptions()
                .setHeadless(false)
                .setSlowMo(800));                
        ctx.browserContext = ctx.browser.newContext();
        ctx.page = ctx.browserContext.newPage();
        ctx.page.onDialog(dialog -> ctx.pendingDialog = dialog);
    }

    @After("@ui")
    public void afterUi(Scenario scenario) {
        try {
            if (scenario.isFailed() && ctx.page != null) {
                try {
                    Path dir = Paths.get("tests", "screenshots");
                    Files.createDirectories(dir);
                    ctx.page.screenshot(new Page.ScreenshotOptions()
                            .setPath(dir.resolve(Instant.now().toEpochMilli() + "-FAILED.png"))
                            .setFullPage(true));
                } catch (Exception ignored) {}
            }
        } finally {
            if (ctx.browserContext != null) ctx.browserContext.close();
            if (ctx.browser != null) ctx.browser.close();
            if (ctx.playwright != null) ctx.playwright.close();
        }
    }
}
